library/src/commonMain/kotlin/com/badoo/reaktive/observable/Zip.kt

```diff
@@ -0,0 +1,231 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun <T, R> Collection<Observable<T>>.zip(mapper: (List<T>) -> R): Observable<R> =
+    observableByEmitter { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+        val lock = newLock()
+        val values = List<MutableList<T>>(size) { mutableListOf() }
+        val readyValues = mutableListOf<T>()
+
+        forEachIndexed { index, source ->
+            source.subscribeSafe(
+                object : ObservableObserver<T> {
+                    override fun onSubscribe(disposable: Disposable) {
+                        disposables += disposable
+                    }
+
+                    override fun onNext(value: T) {
+                        lock.synchronized {
+                            values[index].add(value)
+
+                            if (values.all(List<*>::isNotEmpty)) {
+                                values.forEach { queue ->
+                                    readyValues.add(queue.removeAt(0))
+                                }
+
+                                mapper(readyValues)
+                                    .also {
+                                        readyValues.clear()
+                                        emitter.onNext(it)
+                                    }
+                            }
+                        }
+                    }
+
+                    override fun onComplete() {
+                        lock.synchronized {
+                            emitter.onComplete()
+                        }
+                    }
+
+                    override fun onError(error: Throwable) {
+                        lock.synchronized {
+                            emitter.onError(error)
+                        }
+                    }
+                }
+            )
+        }
+    }
+
+fun <T, R> zip(vararg sources: Observable<T>, mapper: (List<T>) -> R): Observable<R> =
+    sources.toList().zip(mapper)
+
+fun <T1, T2, R> zip(
+    source1: Observable<T1>,
+    source2: Observable<T2>,
+    mapper: (T1, T2) -> R
+): Observable<R> =
```

Comment:
Could use reified to avoid unchecked cast.

Reply:
See my comment for ConbineLatest